### PR TITLE
HDDS-11914. Snapshot diff should not filter SST Files based by reading SST file reader

### PR DESCRIPTION
**Pull request by swamirishi**
## What changes were proposed in this pull request?
Snapshot diff should not filter SST Files based by reading SST file reader and instead should rely on rocksdb's liveMetadataFile which also includes tombstone entries. 

When creating dag compaction log as well, key ranges should also look at liveMetadataFile.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11914

## How was this patch tested?
Existing Unit tests and adding more

### CODE REVIEW

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java [hunk 1]
```diff
@@ -73,44 +71,38 @@ public static String constructBucketKey(String keyName) {
...
-      return false;
-    } catch (RocksDBException e) {
-      LOG.error("Failed to read SST File ", e);
-      throw new IOException(e);
+      CompactionNode compactionNode = preExistingCompactionNodes.get(filename);
+      if (compactionNode == null) {
+        compactionNode = new CompactionNode(new CompactionFileInfo.Builder(filename)
+            .setValues(liveFileMetaDataMap.get(filename)).build());
+      }
+      if (RocksDBCheckpointDiffer.shouldSkipNode(compactionNode, tableToPrefixMap)) {
```
Comment (hemantk-12) [inline hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:103]: nit: move `shouldSkipNode` from  RocksDBCheckpointDiffer to RocksDiffUtils when working on HDDS-11908

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 2]
```diff
@@ -1155,21 +1157,16 @@ Set<String> getDeltaFiles(OmSnapshot fromSnapshot,
...
-                  .getDb().getManagedRocksDb(),
-              tablesToLookUp);
-      Set<String> toSnapshotFiles =
-          RdbUtil.getSSTFilesForComparison(
-              ((RDBStore)toSnapshot.getMetadataManager().getStore()).getDb()
-                  .getManagedRocksDb(),
-              tablesToLookUp);
+      ManagedRocksDB fromDB = ((RDBStore)fromSnapshot.getMetadataManager().getStore())
+          .getDb().getManagedRocksDb();
+      ManagedRocksDB toDB = ((RDBStore)fromSnapshot.getMetadataManager().getStore())
```
Comment (hemantk-12) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:1162]: Typo `toSnapshot`
```suggestion
      ManagedRocksDB toDB = ((RDBStore) toSnapshot.getMetadataManager().getStore())
          .getDb().getManagedRocksDb();
```

File: hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksDB.java [hunk 3]
```diff
@@ -31,13 +32,21 @@
...
 
 /**
  * Managed {@link RocksDB}.
  */
 public class ManagedRocksDB extends ManagedObject<RocksDB> {
   public static final Class<RocksDB> ORIGINAL_CLASS = RocksDB.class;
   public static final int NOT_FOUND = RocksDB.NOT_FOUND;
+  /**
+   * SST file extension. Must be lower case.
+   * Used to trim the file extension when writing compaction entries to the log
```
Comment (hemantk-12) [inline hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksDB.java:46]: nit: please update this comment.

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java [hunk 4]
```diff
@@ -144,14 +143,8 @@ public class RocksDBCheckpointDiffer implements AutoCloseable,
...
 
-  /**
-   * SST file extension. Must be lower case.
-   * Used to trim the file extension when writing compaction entries to the log
-   * to save space.
-   */
-  static final String SST_FILE_EXTENSION = ".sst";
   public static final int SST_FILE_EXTENSION_LENGTH =
-      SST_FILE_EXTENSION.length();
+      ManagedRocksDB.SST_FILE_EXTENSION.length();
```
Comment (hemantk-12) [inline hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java:147]: nit: static import `SST_FILE_EXTENSION`.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 5]
```diff
@@ -1155,21 +1157,16 @@ Set<String> getDeltaFiles(OmSnapshot fromSnapshot,
...
+          .getDb().getManagedRocksDb();
+      Set<String> fromSnapshotFiles = RdbUtil.getSSTFilesForComparison(fromDB, tablesToLookUp);
```
Comment (hemantk-12) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:1164]: Please use the helper function `SnapshotDiffManager#getSSTFileListForSnapshot()`;

File: hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksDB.java [hunk 6]
```diff
@@ -102,4 +111,15 @@ public void deleteFile(LiveFileMetaData fileToBeDeleted)
     File file = new File(fileToBeDeleted.path(), fileToBeDeleted.fileName());
     ManagedRocksObjectUtils.waitForFileDelete(file, Duration.ofSeconds(60));
   }
+
+  public static Map<String, LiveFileMetaData> getLiveMetadataForSSTFiles(RocksDB db) {
+    return db.getLiveFilesMetaData().stream()
+        .filter(liveFileMetaData -> liveFileMetaData.fileName().endsWith(SST_FILE_EXTENSION)).collect(
```
Comment (hemantk-12) [inline hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksDB.java:117]: `db.getLiveFilesMetaData()` only returns table files, why do we need to filter it?
RocksDB documentation:
```
  /**
   * Gets a list of all table files metadata.
   *
   * @return table files metadata.
   */
  public List<LiveFileMetaData> getLiveFilesMetaData() {
    return Arrays.asList(getLiveFilesMetaData(nativeHandle_));
  }
  ```

Comment (swamirishi) [reply hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksDB.java:117]: It could also return manifest and log files

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java (hunk 1 above)
Comment (swamirishi) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:103]: will do it

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java (hunk 2 above)
Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java]: done

File: hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksDB.java (hunk 3 above)
Comment (swamirishi) [reply hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksDB.java]: reverted this change. Not required

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java (hunk 4 above)
Comment (swamirishi) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java]: We don't need it anymore

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java (hunk 5 above)
Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java]: done

File: hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksDB.java (hunk 6 above)
Comment (swamirishi) [reply hadoop-hdds/managed-rocksdb/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksDB.java]: removed the filtering logic. I thought we are also returning log file as part of this.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java (hunk 2 above)
Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java]: Added a test case as well for this.

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java [hunk 7]
```diff
@@ -73,44 +71,38 @@ public static String constructBucketKey(String keyName) {
...
+  public static void filterRelevantSstFiles(Set<String> inputFiles,
+                                            Map<String, String> tableToPrefixMap,
+                                            Map<String, CompactionNode> preExistingCompactionNodes,
+                                            ManagedRocksDB... dbs) {
+    Map<String, LiveFileMetaData> liveFileMetaDataMap = new HashMap<>();
+    int dbIdx = 0;
+    for (Iterator<String> fileIterator =
+         inputFiles.iterator(); fileIterator.hasNext();) {
+      String filename = FilenameUtils.getBaseName(fileIterator.next());
+      while (!preExistingCompactionNodes.containsKey(filename) && !liveFileMetaDataMap.containsKey(filename)
```
Comment (hemantk-12) [inline hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:91]: 1. I don't think we need nested while loop.
2. Can it be moved outside the for loop?

Comment (swamirishi) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:91]: We don't have to read all rocksdb. We can minimise number of rocksdb to be read.

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java [hunk 8]
```diff
@@ -580,8 +580,8 @@ private void createLink(Path link, Path source) {
   private long getSSTFileSummary(String filename)
       throws RocksDBException, FileNotFoundException {
 
-    if (!filename.endsWith(SST_FILE_EXTENSION)) {
-      filename += SST_FILE_EXTENSION;
+    if (!filename.endsWith(ManagedRocksDB.SST_FILE_EXTENSION)) {
```
[the remark this answers is not in the document]
Comment (swamirishi) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java]: done

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java (hunk 7 above)
Comment (hemantk-12) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:91]: Once it gets inside the outer while loop, inside while loop will load liveFileMetaData from both snapshots.

Comment (swamirishi) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:91]: No it won't. It will only load the second snapshot if the file is not present in the first one or the compaction log

Comment (swamirishi) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:91]: Oh ok the second one was supposed to be an if

Comment (swamirishi) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDiffUtils.java:91]: done

**Review by hemantk-12 — COMMENTED** [review]
Thanks @swamirishi for the patch.

LGTM.

**Review by prashantpogde — APPROVED** [review]
Changes look good to me. Thank you for making tehse changes @swamirishi

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 9]
```diff
@@ -1032,8 +1032,10 @@ private void getDeltaFilesAndDiffKeysToObjectIdToKeyMap(
...
     // TODO: [SNAPSHOT] Update Rocksdb SSTFileIterator to read tombstone
     if (skipNativeDiff || !isNativeLibsLoaded) {
-      deltaFiles.addAll(getSSTFileListForSnapshot(fromSnapshot,
-          tablesToLookUp));
+      Set<String> inputFiles = getSSTFileListForSnapshot(fromSnapshot, tablesToLookUp);
+      ManagedRocksDB fromDB = ((RDBStore)fromSnapshot.getMetadataManager().getStore()).getDb().getManagedRocksDb();
+      RocksDiffUtils.filterRelevantSstFiles(inputFiles, tablePrefixes, fromDB);
+      deltaFiles.addAll(inputFiles);
     }
     addToObjectIdMap(fsTable, tsTable, deltaFiles,
```
Comment (hemantk-12) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:1040]: nit: can you please log `deltaFiles` at DEBUG level?

**Review by hemantk-12 — APPROVED** [review]
Thanks @swamirishi for the fix.

LGTM.

**Comment by swamirishi** [comment]
Thanks @hemantk-12 @prashantpogde for reviewing the patch

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java [hunk 10]
```diff
@@ -825,18 +834,16 @@ private String getSSTFullPath(String sstFilenameWithoutExtension,
...
    *               "/path/to/sstBackupDir/000060.sst"]
    */
-  public synchronized Optional<List<String>> getSSTDiffListWithFullPath(
-      DifferSnapshotInfo src,
-      DifferSnapshotInfo dest,
-      String sstFilesDirForSnapDiffJob
-  ) throws IOException {
+  public synchronized Optional<List<String>> getSSTDiffListWithFullPath(DifferSnapshotInfo src,
+                                                                        DifferSnapshotInfo dest,
+                                                                        String sstFilesDirForSnapDiffJob) {
```
Comment (adoroszlai) [inline hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java:839]: nit: Indenting parameters like this triggers unnecessary change whenever visibility / return type / method name is updated.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDiffManager.java [hunk 11]
```diff
@@ -69,6 +69,7 @@
 import org.apache.ratis.util.TimeDuration;
 import jakarta.annotation.Nonnull;
 import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDiffManager.java:72]: nit: please use static import for assertions and mocks (see HDDS-9961).
